New XC Asset Registration Process (#1013)
moonbeam-foundation/moonbeam-docs: builders/interoperability/xcm/xc-registration/assets.md

To create a forum post on the [Moonbeam Community Forum](https://forum.moonbeam.network){target=\_blank}, you'll need to make sure that you're adding the post to the correct category and adding relevant content. For general guidelines and a template to follow, please refer to the [Moonbeam Community Forum Templates for XCM Integrations](/builders/interoperability/xcm/xc-registration/forum-templates/#){target=\_blank} page.

### Calculate Relative Price {: #calculate-relative-price }

> An asset's `relativePrice` refers to a `u128` value that indicates how many units of said asset (in its smallest denomination) equate to one unit**—i.e., `1 × 10^18 wei`—of the native token (GLMR or MOVR). This helps determine how much of your asset to use for fees initially quoted in the native token, particularly in cross-chain messaging (XCM).

Reviewer: Capitalize Wei

You can use the following script (also available as part of [XCM-tools](https://github.com/Moonsong-Labs/xcm-tools){target=\_blank} ) to calculate the correct `relativePrice` value for your asset.

??? code "Calculate Relative Price"
    ```typescript
    --8<-- 'code/builders/interoperability/xcm/xc-registration/assets/calculate-relative-price.ts'
    ```

Only three parameters are required to calculate the relative price of an asset: 

- Asset Price (USD): A positive number representing how much 1 unit (in human-readable form) of your asset costs in USD
- Asset Decimals: The number of decimal places your asset uses. For example, if your token has 12 decimals, specify 12
- Network: Either GLMR (Moonbeam) or MOVR (Moonriver). This should correspond to the network that you're registering the asset on, and this determines which native token’s USD price the script will fetch from CoinGecko

First, ensure that you've installed the required dependencies by running:

```bash
yarn
```

Execute the script, making sure to provide the USD price of the asset you're registering, the number of decimals it has, and the network you're registering the asset on (either GLMR or MOVR):

```bash

> yarn calculate-relative-price <INSERT_ASSET_PRICE> <INSERT_DECIMALS> GLMR

Reviewer: ```suggestion
yarn calculate-relative-price INSERT_ASSET_PRICE INSERT_DECIMALS GLMR
```

Upon successful execution, the script prints the computed `relativePrice` as a `BigInt`. This value represents the scaled ratio between the asset’s USD price and the native token’s USD price, multiplied up to 18 decimals. You can then use this result in on-chain asset registration or fee calculation scenarios—especially where a `u128` 18-decimal format is required.

For additional info, usage details, or to see an example in action, you can invoke the help command by running: 

```bash
yarn calculate-relative-price --help
```

### Generate the Encoded Calldata for the Asset Registration {: #generate-encoded-calldata-for-asset-registration }

Submitting a governance proposal on Moonbeam requires two steps: first, submit a preimage that defines the actions to be executed, then use that preimage to submit the proposal. For more details, see the [Governance on Moonbeam](/learn/features/governance/){target=\_blank} page. To submit a preimage for asset registration, you'll need the encoded calldata for both the `evmForeignAssets.createForeignAsset` and `xcmWeightTrader.addAsset` extrinsics. An existing asset's price can be updated with `xcmWeightTrader.editAsset`. 

> Proposals must be submitted via the `FastGeneralAdmin` track. A channel must be established before an asset can be registered. To get the encoded calldata for the `evmForeignAssets.createForeignAsset` extrinsic, you will need to provide the following arguments:

Reviewer: ```suggestion
Proposals must be submitted via the Fast General Admin track. A channel must be established before an asset can be registered. To get the encoded calldata for the `evmForeignAssets.createForeignAsset` extrinsic, you will need to provide the following arguments:
```

> - `assetId` - unique identifier of the asset, generated from the [`calculate-external-asset-info.ts`](https://github.com/Moonsong-Labs/xcm-tools/blob/main/scripts/calculate-external-asset-info.ts){target=\_blank} script

Reviewer: ```suggestion
- **`assetId`** - unique identifier of the asset, generated from the [`calculate-external-asset-info.ts`](https://github.com/Moonsong-Labs/xcm-tools/blob/main/scripts/calculate-external-asset-info.ts){target=\_blank} script
```

Standards have evolved, so we should be putting all of these items on the left in bold. Can you please update the rest of the items in this list?

- `xcmLocation` - the multilocation of the asset relative to Moonbeam 
- `decimals` - the number of decimals of the asset
- `symbol`  - the symbol of the asset. Remember that "xc" should be prepended to the symbol to indicate the asset is an XCM-enabled asset

> - `name` - The asset name

Reviewer: ```suggestion
- `name` - the asset name
```

Using the above information, you can generate the encoded call data for the `createForeignAsset` call either via the Polkadot API or on [Polkadot.js Apps](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fwss.api.moonbeam.network#/extrinsics){target=\_blank}.

You can generate this required calldata using the [xcm-asset-registrator script](https://github.com/Moonsong-Labs/xcm-tools/blob/main/scripts/xcm-asset-registrator.ts){target=\_blank} as follows:

```bash
yarn register-asset --w wss://wss.api.moonbeam.network  \

> --asset 'INSERT_MULTILOCATION' \

Reviewer: ```suggestion
--asset "INSERT_MULTILOCATION" \
```

--symbol "INSERT_ASSET_SYMBOL" \
--decimals INSERT_DECIMALS \
--name "INSERT_ASSET_NAME" \
--relative-price INSERT_RELATIVE_PRICE
```

Upon running the script with the relevant parameters, you'll see output like the following: 

--8<-- 'code/builders/interoperability/xcm/xc-registration/assets/terminal/register-asset.md'

The script will provide the encoded call data for each of the following calls:

- The `registerAsset` call
- The `setRelativePrice` call
- The `batch` call that combines each all of the above

![Overview of the proposal process](/images/builders/interoperability/xcm/xc-registration/assets/assets-3.webp)

### Construct the Add Asset Call

If you've already used the [xcm-asset-registrator script](https://github.com/Moonsong-Labs/xcm-tools/blob/main/scripts/xcm-asset-registrator.ts){target=\_blank} shown above, you can skip this section. This section dives into more detail about how the `xcmWeightTrader.addAsset` call is constructed. To get the encoded calldata for the `xcmWeightTrader.addAsset` extrinsic, you will need to provide the following arguments:

> - `xcmLocation` - the multilocation of the asset relative to Moonbeam 

Reviewer: ```suggestion
- **`xcmLocation`** - the multilocation of the asset relative to Moonbeam 
```

Bold here too please, please check rest of the guide

- `relativePrice` - A numeric value (u128) representing the fraction of the native token’s price that your asset’s price constitutes, scaled to 18 decimals. This value calculates cross-chain fees by determining how many units of the non-native asset are required to cover XCM operation costs. 

Using the above information, you can generate the encoded call data for the `addAsset` call either via the Polkadot API or on [Polkadot.js Apps](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fwss.api.moonbeam.network#/extrinsics){target=\_blank}.

> To create a batch transaction that combines both the `xcmWeightTrader.addAsset` and the `evmForeignAssets.createForeignAsset` calls together, you can use the [Polkadot API's Batch Method](/builders/substrate/libraries/polkadot-js-api/#batching-transactions){target=\_blank}. As mentioned previously, the [XCM Asset Registrator](https://github.com/Moonsong-Labs/xcm-tools/blob/main/scripts/xcm-asset-registrator.ts){target=\_blank} script can help you build and submit the required calls. 

Reviewer: ```suggestion
To create a batch transaction that combines both the `xcmWeightTrader.addAsset` and the `evmForeignAssets.createForeignAsset` calls together, you can use the [Polkadot API's `batch` method](/builders/substrate/libraries/polkadot-js-api/#batching-transactions){target=\_blank}. As mentioned previously, the [XCM asset registrator script](https://github.com/Moonsong-Labs/xcm-tools/blob/main/scripts/xcm-asset-registrator.ts){target=\_blank} can help you build and submit the required calls. 
```

No need to capitalize Batch Method. The only reason why you would want to capitalize when linking is if the title is a 1:1 match.

### Submit the Preimage and Proposal for Asset Registration {: #submit-preimage-proposal }

Your next task is to submit the preimage of your batched call containing both the `xcmWeightTrader.addAsset` and the `evmForeignAssets.createForeignAsset` by following the guidelines in the [Submit a Democracy Proposal Guide](/tokens/governance/proposals/#submitting-a-preimage-of-the-proposal){target=\_blank}.

> For Moonbase Alpha, you do not need to go through governance, as Moonbase Alpha has `sudo` access. Instead, you can provide the output of the batch call data to the Moonbeam team, and the Moonbeam Team can submit the call with `sudo`. This will be a faster and easier process than going through governance. However, you may still wish to go through governance on Moonbase Alpha to prepare for Moonbeam's governance process.

Reviewer: ```suggestion
You do not need to go through governance for Moonbase Alpha, as Moonbase Alpha has sudo access. Instead, you can provide the output of the batch call data to the Moonbeam team, and they can submit the call with sudo. This will be a faster and easier process than going through governance. However, you may still wish to go through governance on Moonbase Alpha to prepare for Moonbeam's governance process.
```

After submitting the preimage, you can submit the proposal by following the guidelines in the [Submitting a Proposal](/tokens/governance/proposals/#submitting-a-proposal-v2){target=\_blank} section.

> If you prefer the script method and you're comfortable working with the scripts in the XCM tools repo, you can use the [Generic Call Proposer](https://github.com/Moonsong-Labs/xcm-tools/blob/main/scripts/generic-call-proposer.ts){target=\_blank} by passing in the requisite calls, including the acceptance and proposal of the XCM Channel, and the asset registration. The [Generic Call Proposer](https://github.com/Moonsong-Labs/xcm-tools/blob/main/scripts/generic-call-proposer.ts){target=\_blank} can help you assemble the multiple requisite calls as follows:

Reviewer: ```suggestion
If you prefer the script method and you're comfortable working with the scripts in the XCM tools repo, you can use the [generic call proposer](https://github.com/Moonsong-Labs/xcm-tools/blob/main/scripts/generic-call-proposer.ts){target=\_blank} by passing in the requisite calls, including the acceptance and proposal of the XCM Channel, and the asset registration. The [generic call proposer](https://github.com/Moonsong-Labs/xcm-tools/blob/main/scripts/generic-call-proposer.ts){target=\_blank} can help you assemble the multiple requisite calls as follows:
```

### Create a Proposal to Register an Asset {: #create-a-proposal }

To register an asset native to another chain on Moonbeam, you'll need to submit a governance proposal that will call the `evmForeignAssets.createForeignAsset` extrinsic. Additionally, you'll need to register the asset's price (formerly known as units per second) through the `xcmWeightTrader.addAsset` extrinsic so that XCM fees can be properly calculated. An existing asset's price can be updated with `xcmWeightTrader.editAsset`. 

To get started, you'll need to collect some information about the asset:

- `assetId`: unique identifier of the asset
- `xcmLocation`: a scale encoded xcm v4 multilocation of asset reserve relative to Moonbeam
- The asset name
- The asset symbol. You'll need to prepend "xc" to the asset symbol to indicate that the asset is an XCM-enabled asset
- The number of decimals the asset has

> With this information in hand, you can prepare a governance proposal to register a foreign asset. Foreign asset registration proposals should be submitted through the `FastGeneralAdmin` track.

Reviewer 2: Asset  registration needs to be done via General Admin.

THe only thing that can be done in FastGeneralAdmin is opening the channel -> https://github.com/moonbeam-foundation/moonbeam/blob/master/runtime/moonbeam/src/xcm_config.rs#L641

From Moonbase -> https://github.com/moonbeam-foundation/moonbeam/blob/527a930e6f02522edc26047c36518b3b712705cf/runtime/moonbase/src/xcm_config.rs#L699-L705

![Overview of the proposal process](/images/builders/interoperability/xcm/xc-registration/assets/assets-3.webp)

### Generate the Encoded Calldata for the Asset Registration {: #generate-encoded-calldata-for-asset-registration }

If you're not familiar with the governance system on Moonbeam, you can find out more information on the [Governance on Moonbeam](/learn/features/governance/){target=\_blank} page. With any governance proposal on Moonbeam, you'll need to submit a preimage, which defines the actions to be executed, and then use the preimage to submit a proposal.

To submit a preimage, you'll need to get the encoded calldata for each extrinsic that you want to execute. As previously mentioned, you'll use the `evmForeignAssets.createForeignAsset` and the `xcmWeightTrader.addAsset` extrinsics.

Proposals must be submitted via the `FastGeneralAdmin` track. If you're opening a channel and registering an asset and you'll want to wait until the channel is established prior to attempting to register the asset. 

To get the encoded calldata for the `evmForeignAssets.createForeignAsset` extrinsic, you will need to provide the following arguments:

> - `assetId` - unique identifier of the asset

Reviewer 2: We never explain how to get the assetID, actually, because the user hast o provide the multilocation we can modify `xcm-tools` to create this for you

- `xcmLocation` - the multilocation of the asset relative to Moonbeam 
- `decimals` - the number of decimals of the asset
- `symbol`  - the symbol of the asset. **Remember that "xc" should be prepended to the symbol to indicate the asset is an XCM-enabled asset**
- `name` - The asset name

You can use the script below to generate the encoded call data for the `createForeignAsset` call. Remember to replace the example parameter values shown in the script with the ones relevant to your asset. 

???+ code "Get encoded call data for createForeignAsset"
    ```typescript
    --8<-- 'code/builders/interoperability/xcm/xc-registration/assets/create-foreign-asset.js'
    ```

The resulting output will be something like:

--8<-- 'code/builders/interoperability/xcm/xc-registration/assets/terminal/createForeignAsset.md'

To get the encoded calldata for the `xcmWeightTrader.addAsset` extrinsic, you will need to provide the following arguments:

- `xcmLocation` - the multilocation of the asset relative to Moonbeam 

> - `relativePrice` - the cost of an asset in terms of weight, used to determine how much of the asset is required to cover the fees for cross-chain (XCM) operations. It is calculated by comparing the asset's value to the network's native token in terms of the weight-to-fee conversion

Reviewer 2: How do you get this? We need to provide a practical example or make `xcm-tools` calculate this for you!

Author: Added a script on XCM-tools :)

- The asset name
- The asset symbol. You'll need to prepend "xc" to the asset symbol to indicate that the asset is an XCM-enabled asset
- The number of decimals the asset has

With this information in hand, you can prepare a governance proposal to register a foreign asset. Foreign asset registration proposals should be submitted through the `GeneralAdmin` track.

![Overview of the proposal process](/images/builders/interoperability/xcm/xc-registration/assets/assets-3.webp)

### Calculate Relative Price {: #calculate-relative-price }

An asset's `relativePrice` refers to a `u128` value that indicates how many units of said asset (in its smallest denomination) equate to one unit**—i.e., `1 × 10^18 wei`—of the native token (GLMR or MOVR). This helps determine how much of your asset to use for fees initially quoted in the native token, particularly in cross-chain messaging (XCM).

You can use the following script (also available as part of [XCM-tools](https://github.com/Moonsong-Labs/xcm-tools){target=\_blank} ) to calculate the correct `relativePrice` value for your asset.

??? code "Calculate Relative Price"
    ```typescript

>     --8<-- 'code/builders/interoperability/xcm/xc-registration/assets/calculate-relative-price.ts'

Author: It appears that Mkdocs does not support this behavior but we can look into optimizing this in the future perhaps with some kind of external plugin.

### Generate the Encoded Calldata for the Asset Registration {: #generate-encoded-calldata-for-asset-registration }

Submitting a governance proposal on Moonbeam requires two steps: first, submit a preimage that defines the actions to be executed, then use that preimage to submit the proposal. For more details, see the [Governance on Moonbeam](/learn/features/governance/){target=\_blank} page. To submit a preimage for asset registration, you'll need the encoded calldata for both the `evmForeignAssets.createForeignAsset` and `xcmWeightTrader.addAsset` extrinsics.

Proposals must be submitted via the `GeneralAdmin` track. A channel must be established before an asset can be registered. To get the encoded calldata for the `evmForeignAssets.createForeignAsset` extrinsic, you will need to provide the following arguments:

> - `assetId` - unique identifier of the asset, generated from the [`calculate-external-asset-info.ts`](https://github.com/Moonsong-Labs/xcm-tools/blob/main/scripts/calculate-external-asset-info.ts){target=\_blank} script

Author: Done!

### Submit the Preimage and Proposal for Asset Registration {: #submit-preimage-proposal }

Your next task is to submit the preimage of your batched call containing both the `xcmWeightTrader.addAsset` and the `evmForeignAssets.createForeignAsset` by following the guidelines in the [Submit a Democracy Proposal Guide](/tokens/governance/proposals/#submitting-a-preimage-of-the-proposal){target=\_blank}.

For Moonbase Alpha, you do not need to go through governance, as Moonbase Alpha has `sudo` access. Instead, you can provide the output of the batch call data to the Moonbeam team, and the Moonbeam Team can submit the call with `sudo`. This will be a faster and easier process than going through governance. However, you may still wish to go through governance on Moonbase Alpha to prepare for Moonbeam's governance process.

After submitting the preimage, you can submit the proposal by following the guidelines in the [Submitting a Proposal](/tokens/governance/proposals/#submitting-a-proposal-v2){target=\_blank} section.

> If you prefer the script method and you're comfortable working with the scripts in the XCM tools repo, you can use the [Generic Call Proposer](https://github.com/Moonsong-Labs/xcm-tools/blob/main/scripts/generic-call-proposer.ts){target=\_blank} by passing in the requisite calls, including the acceptance and proposal of the XCM Channel, and the asset registration. 

Reviewer 2: Like:

yarn generic call --data "CALLDATA_INCOMING_XCM_CHANNEL" --data "CALLDATA_OUTGOING_XCM_CHANNL" --data "CALLDATA_BATCH_ASSET_REGISTRATION"

> If you prefer the script method and you're comfortable working with the scripts in the XCM tools repo, you can use the [Generic Call Proposer](https://github.com/Moonsong-Labs/xcm-tools/blob/main/scripts/generic-call-proposer.ts){target=\_blank} by passing in the requisite calls, including the acceptance and proposal of the XCM Channel, and the asset registration. 

Author: Done!

### Create a Proposal to Register an Asset {: #create-a-proposal }

To register an asset native to another chain on Moonbeam, you'll need to submit a governance proposal that will call the `evmForeignAssets.createForeignAsset` extrinsic. Additionally, you'll need to register the asset's price (formerly known as units per second) through the `xcmWeightTrader.addAsset` extrinsic so that XCM fees can be properly calculated. An existing asset's price can be updated with `xcmWeightTrader.editAsset`. 

To get started, you'll need to collect some information about the asset:

- `assetId`: unique identifier of the asset. This can be any unique arbitrary integer, but it's recommended that you create this using the [`calculate-external-asset-info.ts`](https://github.com/Moonsong-Labs/xcm-tools/blob/main/scripts/calculate-external-asset-info.ts){target=\_blank} script. This provides a standardized way to generate a unique `assetId`

> - `xcmLocation`: a scale encoded xcm v4 multilocation of asset reserve relative to Moonbeam

Reviewer 3: ```suggestion
- `xcmLocation` - a scale encoded xcm v4 multilocation of asset reserve relative to Moonbeam
```

    }
    ```

Since local XC-20s are ERC-20s on Moonbeam, there are no deposits required to create an ERC-20 on Moonbeam. However, deposits may be required to register the asset on another parachain. Please consult with the parachain team you wish to register the asset with for more information.

## Managing XC Assets 

After completing the [registration process](#introduction) for an XC asset, you may need to periodically update asset details, such as the XCM multilocation details or asset price. This section will cover these topics.

### Updating Foreign Asset XCM Location {: #updating-foreign-asset-xcm-location }

> You can update the multilocation of an asset with the `evmForeignAssets.changeXcmLocation` call, which takes as parameters the `assetId` and the new multilocation. You'll need to raise a [governance proposal](/tokens/governance/proposals/) and submit the update under the `GeneralAdmin` track. If you're testing in Moonbase Alpha, you can ask the Moonbeam Team to submit the extrinsic using Sudo to speed up the process. You can also submit the requisite governance proposal on Moonbase Alpha. 

Reviewer: ```suggestion
You can update the multilocation of an asset with the `evmForeignAssets.changeXcmLocation` call, which takes as parameters the `assetId` and the new multilocation. You'll need to raise a [governance proposal](/tokens/governance/proposals/) and submit the update under the General Admin track. If you're testing in Moonbase Alpha, you can ask the Moonbeam Team to submit the extrinsic using Sudo to speed up the process. You can also submit the requisite governance proposal on Moonbase Alpha. 
```
